src/Processors/Merges/Algorithms/FinishAggregatingInOrderAlgorithm.h

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include <Processors/Merges/Algorithms/IMergingAlgorithm.h>
+#include <Processors/Merges/Algorithms/MergedData.h>
+#include <Core/SortDescription.h>
+#include <Core/Block.h>
+
+namespace DB
+{
+
+struct AggregatingTransformParams;
+using AggregatingTransformParamsPtr = std::shared_ptr<AggregatingTransformParams>;
+
+/**
+ * The second step of aggregation in order of sorting key.
+ * The transform receives k inputs with partially aggregated data,
+ * sorted by group by key (prefix of sorting key).
+ * Then it merges aggregated data from inputs by the following algorithm:
+ *  - At each step find the smallest value X of the sorting key among last rows of current blocks of inputs.
+ *    Since the data is sorted in order of sorting key and has no duplicates (because of aggregation),
+ *    X will never appear later in any of input streams.
+ *  - Aggregate all rows in current blocks of inputs up to the upper_bound of X using
+ *    regular hash table algorithm (Aggregator::mergeBlock).
```

Comment:
I don't understand this part -- if the input data is sorted by the `GROUP BY` key, and we cut the parts of input blocks with same keys, and pass them to `mergeBlock` -- the hash table is always going to have only one element? Would it make sense to just use one aggregation state?

Reply:
There can be duplicates among all streams for one range. The hash table at one step will contain all keys <= X from all blocks, but not one element. There are already algorithm (`AggregatingSortedAlgorithm`), that merges and aggregates sorted data and uses one aggregation state. It is a simple k-way merge algorithm and it makes `O(n*log(k))` comparisons, where n -- number of rows, k -- number of parts. Algorithm from this PR makes `O(n + k * log(n)) ` operations, n -- for hash table, `k * log(n)` -- for finding positions in blocks. It is better than `O(n*log(k))`, when k is not too big and not too small (about 100-1000).

Reply:
Thanks, I think I understand now.

Reply:
Maybe we should just copy-paste this explanation into the comment as well? It mentions the other algorithm and discusses the complexity, I think this is a useful comment.